org.lflang/src/org/lflang/generator/ts/TsGenerator.kt

```diff
@@ -0,0 +1,436 @@
+    fun getTargetTypeW(t: Type): String = getTargetType(t)
+
+    fun getInitializerListW(state: StateVar): List<String> = getInitializerList(state)
+    fun getInitializerListW(param: Parameter): List<String> = getInitializerList(param)
+    fun getInitializerListW(param: Parameter, i: Instantiation): List<String> =
+        getInitializerList(param, i)
+    fun generateVarRefW(reference: VarRef): String =generateVarRef(reference)
+
+    /** Generate TypeScript code from the Lingua Franca model contained by the
+     *  specified resource. This is the main entry point for code
+     *  generation.
+     *  @param resource The resource containing the source code.
+     *  @param fsa The file system access (used to write the result).
+     *  @param context FIXME: Undocumented argument. No idea what this is.
+     */
+    override fun doGenerate(resource: Resource, fsa: IFileSystemAccess2,
+                            context: IGeneratorContext) {
+        super.doGenerate(resource, fsa, context)
+
+        // stop if there are any errors found in the program by doGenerate() in GeneratorBase
+        if (errorsOccurred()) return
+
+        // abort if there is no main reactor
+        if (mainDef == null) {
+            println("WARNING: The given Lingua Franca program does not define a main reactor. Therefore, no code was generated.")
+            return
+        }
+
+        fileConfig.deleteDirectory(fileConfig.srcGenPath)
+        for (runtimeFile in RUNTIME_FILES) {
+            copyFileFromClassPath(
+                "/lib/TS/reactor-ts/src/core/$runtimeFile",
+                tsFileConfig.tsCoreGenPath().resolve(runtimeFile).toString())
+        }
+
+        /**
+         * Check whether configuration files are present in the same directory
+         * as the source file. For those that are missing, install a default
+         * If the given filename is not present in the same directory as the source
+         * file, copy a default version of it from /lib/TS/.
+         */
+        for (configFile in CONFIG_FILES) {
+            val configFileDest = fileConfig.srcGenPath.resolve(configFile).toString()
+            val configFileInSrc = fileConfig.srcPath.resolve(configFile).toString()
+            if (fsa.isFile(configFileInSrc)) {
+                // TODO(hokeun): Check if this logic is still necessary.
+                println("Copying '" + configFile + "' from " + fileConfig.srcPath)
+                copyFileFromClassPath(configFileInSrc, configFileDest)
+            } else {
+                println(
+                    "No '" + configFile + "' exists in " + fileConfig.srcPath +
+                            ". Using default configuration."
+                )
+                copyFileFromClassPath("/lib/TS/$configFile", configFileDest)
+            }
+        }
+
+        for (federate in federates) {
+            var tsFileName = fileConfig.name
+            if (isFederated) {
+                tsFileName += '_' + federate.name
```

Comment:
There is a [FedFileConfig](https://github.com/icyphy/lingua-franca/blob/ca8602ad799de719d111586c292fc880eecbe867/org.lflang/src/org/lflang/federated/FedFileConfig.java) class that might be useful here. [Here](https://github.com/icyphy/lingua-franca/blob/ca8602ad799de719d111586c292fc880eecbe867/org.lflang/src/org/lflang/generator/c/CGenerator.xtend#L462) is an example of its use in the C target. The behavior in C target is to put each federate into its own folder.

Reply:
Thanks for the suggestion. I believe the federated execution is not supported yet in TypeScript and I'm planning to fix that later. I added a TODO to follow up. Would this be ok with you for now?